### PR TITLE
Fix/Extend UARTE

### LINKED ISSUE #42: UARTE Pins are wrong
Right now, RXD and CTS of Uarte::Pins are of type output. This is wrong.

https://github.com/nrf-rs/nrf52-hal/blob/master/nrf52-hal-common/src/uarte.rs#L172-L177

From the datasheet:

| UARTE signal | UARTE pin | Direction | Output value |
| :--- | :--- | :--- | :--- |
| RXD | As specified in PSEL.RXD | Input | N/A |
| CTS | As specified in PSEL.CTS | Input | N/A |
| RTS | As specified in PSEL.RTS | Output | 1 |
| TXD | As specified in PSEL.TXD | Output | 1 |

### PR DESCRIPTION
**Pull request by jamesmunns**
Closes #42, also adds RX functionality with optional timeout to Uarte.

@hannobraun I'd be interested to hear your opinion on this approach for receiving with a fixed timeout.

### CODE REVIEW

File: nrf52-hal-common/src/timer.rs [hunk 1]
```diff
@@ -146,4 +146,9 @@ impl<T> Timer<T> where T: TimerExt {
 
         Ok(())
     }
+
+    pub fn delay(&mut self, cycles: u32) {
+        self.start(cycles);
+        block!(self.wait()).unwrap();
```
Comment (jamesmunns) [inline nrf52-hal-common/src/timer.rs]: TODO: Should this result be returned?

File: nrf52-hal-common/src/uarte.rs [hunk 2]
```diff
@@ -162,6 +164,130 @@ impl<T> Uarte<T> where T: UarteExt {
...
+        if self.0.rxd.amount.read().bits() != rx_buffer.len() as u32 {
+            return Err(Error::Receive);
+        }
+
+        Ok(())
+    }
+
+    /// Read via UARTE
+    ///
+    /// This method fills all bytes in `rx_buffer`, and blocks
```
Comment (jamesmunns) [inline nrf52-hal-common/src/uarte.rs]: TODO: Fix this comment

File: nrf52-hal-common/src/uarte.rs [hunk 3]
```diff
@@ -162,6 +164,130 @@ impl<T> Uarte<T> where T: UarteExt {
...
+        }
+
+        // Cleanup, even in the error case
+        self.finalize_read();
+
+        let bytes_read = self.0.rxd.amount.read().bits() as usize;
+
+        if timeout_occured {
+            return Err(Error::Timeout(bytes_read));
+        }
```
Comment (jamesmunns) [inline nrf52-hal-common/src/uarte.rs]: TODO: Do we need to explicitly stop/clear the timer here? Probably only if interrupts are enabled...

File: nrf52-hal-common/src/timer.rs (hunk 1 above)
Comment (Yatekii) [reply nrf52-hal-common/src/timer.rs]: I don't think there is any use to this, no?

Comment (jamesmunns) [reply nrf52-hal-common/src/timer.rs]: I think the unwrap will only occur if something other than Error::WouldBlock is returned. That isn't possible today, but might be later? I'm not too familiar with the `nb` crate.

Comment (Yatekii) [reply nrf52-hal-common/src/timer.rs]: Well I guess that would need be adressed when there changes anything about this. It makes no sense to have the user always write

`let _  = delay(42);`

imo.

File: nrf52-hal-common/src/uarte.rs [hunk 4]
```diff
@@ -163,6 +164,130 @@ impl<T> Uarte<T> where T: UarteExt {
...
+        self.start_read(rx_buffer)?;
+
+        // Start the timeout timer
+        timer.start(cycles);
+
+        // Wait for transmission to end
+        let mut event_complete = false;
+        let mut timeout_occured = false;
+
+        loop {
```
Comment (hannobraun) [inline nrf52-hal-common/src/uarte.rs]: I think this is fine, but we should probably document that this method involves busy waiting.

**Comment by hannobraun** [comment]
@jamesmunns 
> I'd be interested to hear your opinion on this approach for receiving with a fixed timeout.

Sorry, forgot to answer this in my review comment. I think this is okay. I'm not exactly thrilled about the busy waiting, but I think it's fine for now. In fact, it might be preferable to dealing with an interrupt, as things are currently.

I think we should revisit this once we've reached the magical land of ownable interrupts, or when it becomes clear that's not going to happen :)

**Comment by jamesmunns** [comment]
@hannobraun addressed all open topics. In general, I'd like the "support longer than 255 DMA bytes" issue to be tackled in a separate PR.

Let me know what you think

**Comment by jamesmunns** [comment]
@hannobraun reping for `r+`/merge

**Comment by hannobraun** [comment]
Looks good to me. Thank you, @jamesmunns!

**Pull request merged**